### PR TITLE
4.x : Workspace Refactoring : Create Datasource Config File

### PR DESCRIPTION
**Pull request by deepakrkris**
Write Datasource config when a new datasource is added

- [X] - add datasource handler to connector
- [X] - add method to class Workspace to get datasource config file path
- [X] - add method to class Datasource to get datasource definition
- [X] - add pretest script to create fresh workspace sandbox for testing
- [X] - add method to class WriteOperations to write to a datasource config file

### CODE REVIEW

File: component/tasks.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ class Tasks {
     const workspace = this;
     //Datasource is a self-aware node which adds itself to the Workspace graph
     new DataSource(workspace, id, datasource);
-    cb(null, datasource);
+    WriteOperations.writeDataSourceConfig(workspace, cb);
```
Comment (superkhau) [inline component/tasks.js]: `WriteOperations` should be `writeOperations` as you are not instantiating via `new`.

File: component/workspace-manager.js [hunk 2]
```diff
@@ -10,6 +10,10 @@ const Manager = class Manager {
   constructor() {
   }
   createWorkspace(dir) {
+    if (this.workspace &&
+      this.workspace.getDirectory() === dir) {
```
Comment (superkhau) [inline component/workspace-manager.js]: nitpick: you can fit the above on one line

File: component/workspace.js [hunk 3]
```diff
@@ -36,6 +44,10 @@ class Workspace extends Graph {
     const ds = this.getNode('DataSource', id);
     return ds;
   }
+  getAllDataSource(id) {
```
Comment (superkhau) [inline component/workspace.js]: What is `id` for? And why is it named `getAllDataSource`?

File: connector/data-source-handler.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+class DataSourceHandler {
+  static createDataSource(workspace, id, data, cb) {
+    const create = function(next) {
```
Comment (superkhau) [inline connector/data-source-handler.js]: `function create(next) {`

File: connector/data-source-handler.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      workspace.addDataSource(id, data, function(err) {
+        next(err);
+      });
+    };
+    const callBack = function(err, results) {
```
Comment (superkhau) [inline connector/data-source-handler.js]: `function callBack(err, results) {`

File: package.json [hunk 6]
```diff
@@ -7,6 +7,7 @@
   },
   "scripts": {
     "lint": "eslint .",
+    "pretest": "node test/helpers/pretest.js",
```
Comment (superkhau) [inline package.json]: why pretest? you should use the funcs directly in the "before" hooks of the mocha tests you want to clean up/create the sandbox for.

File: test/helpers/pretest.js [hunk 7]
```diff
@@ -0,0 +1,6 @@
+'use strict';
+const testSupport = require('./test-support');
+
+testSupport.givenEmptySandbox(function(err) {
+  if (err) throw new Error(err);
```
Comment (superkhau) [inline test/helpers/pretest.js]: `if (err) throw err;`

File: package.json (hunk 6 above)
Comment (deepakrkris) [inline package.json]: mocha is not yet implemented

File: connector/data-source-handler.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+class DataSourceHandler {
+  static createDataSource(workspace, id, data, cb) {
+    function create(next) {
+      workspace.addDataSource(id, data, function(err) {
```
Comment (superkhau) [inline connector/data-source-handler.js]: `err, next`?

`next` on the next line is missing a reference

File: package.json (hunk 6 above)
Comment (superkhau) [inline package.json]: I see, can you do the setup as part of the cucumber feature set via setup/teardown hooks: https://github.com/cucumber/cucumber-js/blob/master/docs/support_files/hooks.md#hooks

Comment (deepakrkris) [inline package.json]: interesting ! , I am going to use this for scenario specific setup, the pretest.js just cleans a sandbox folder and creates a fresh folder. It is not scenario specific

File: connector/data-source-handler.js (hunk 8 above)
Comment (deepakrkris) [inline connector/data-source-handler.js]: the tasks are run by async.queue , which does not provide a err object, there is only next ()

File: package.json (hunk 6 above)
Comment (superkhau) [inline package.json]: Try this: https://github.com/cucumber/cucumber-js/issues/261#issuecomment-59885494 for beforeAll/afterAll hooks. I don't like how its not documented well at all yet.

File: connector/data-source-handler.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        next(err);
```
Comment (superkhau) [inline connector/data-source-handler.js]: This is the line that is referencing nothing. Need to clean up the code to make your intents more clear.

Comment (superkhau) [inline connector/data-source-handler.js]: nevermind, it's good now i see it above on line 4. Ignore this comment. ;)

**Pull request merged**